Week2 - Essential + Advanced(Selection 제외)

Larry7939 opened the pull request.

Larry7939 committed:
Update HomeActivity.kt

Larry7939 committed:
add : 상수 파일 추가

ApplicationClass에서 관리하던 사용자 정보 관련 상수를 user_info_const 파일에서 관리

Larry7939 committed:
add : 상수 파일 추가

ApplicationClass에서 관리하던 사용자 정보 관련 상수를 user_info_const 파일에서 관리

Larry7939 committed:
add : home/gallery/search fragment 추가

Larry7939 committed:
chore HomeActivity -> MainActivity 이름 변경

Larry7939 committed:
add : menu폴더 main_bottom_menu 파일 추가

MainActivity의 bottomNavigationView에 들어갈 menu 리소스 파일 추가

Larry7939 committed:
chore : HomeFragment에 BindingFragment 적용

Larry7939 committed:
chore : 기존 기능 삭제(초기 tv,iv init 및 로그아웃)

Larry7939 committed:
chore : 초기 Fragment를 HomeFragment로 설정

Larry7939 committed:
feat : bottomNavigation으로 fragment 전환 기능 구현

Larry7939 committed:
add : ItemView추가

home RecyclerView에서 사용할 profileImg, RepoName, Author로 구성된 ItemView추가

Larry7939 committed:
add : ItemView에 들어갈 데이터 클래스 추가

Larry7939 committed:
add : HomeRecycleAdapter 추가

Larry7939 committed:
feat : recyclerview 구현

HomeFragment와 adapter연결 및
가짜 데이터 전달

Larry7939 committed:
Update HomeFragment.kt

Larry7939 committed:
add : viewType 상수화

Larry7939 committed:
chore : 데이터 클래스 명칭 변경 및 데이터 클래스에 ViewType추가

ViewType을 고려한 명칭 변경

Larry7939 committed:
add : repo title itemview 레이아웃 작성

Larry7939 committed:
feat : Multi-View Type RecyclerView 구현

Larry7939 committed:
feat : scrollToTop 구현

Larry7939 committed:
Update MainActivity.kt

Larry7939 committed:
Update HomeRecyclerAdapter.kt

Larry7939 committed:
chore : HomeRecycler Data 정리

기존의 home 관련 데이터 클래스, 상수를 data폴더에 정리

Larry7939 committed:
chore : ViewHolder간 상속관계 수정

HomeViewHolder를 RepoContentViewHolder와 RepoTitleViewHolder가 상속받도록 하고, onBind에서 사용하는 데이터 타입을 부모 데이터 클래스 타입(HomeRecycleData)으로 통일

Larry7939 committed:
chore : HomeRecyclerAdapter 간소화

Larry7939 committed:
Update HomeRecyclerAdapter.kt

Larry7939 committed:
Update README.md

Larry7939 committed:
add : DiffUtil ItemCallback클래스 새엇ㅇ

기존에 있던 oldTtem과 새로운 newItem을 비교해서, 변화한 item에 대해서만 데이터 갱신

Larry7939 committed:
Merge branch 'seminar2' of https://github.com/IN-SOPT-ANDROID/jihun-park into seminar2

Larry7939 committed:
add : DiffUtill을 활용한 RecyclerView 성능개선

성능 향상 정도를 프로필 GPU렌더링으로 비교함으로써 확인할 수 있었음.

Larry7939 committed:
Update README.md

Larry7939 committed:
feat : GalleryFragment에 RecyclerView추가

youngjinc commented on app/src/main/AndroidManifest.xml:
```diff
@@ -15,21 +15,21 @@
         tools:targetApi="31">
         <activity
-            android:name=".src.login.SignupActivity"
+            android:name=".presentation.login.SignupActivity"
```
presentation 패키지로 변경되었군요~~ 아주 좋습니다 :)

youngjinc commented on app/src/main/java/org/sopt/sample/presentation/MainActivity.kt:
```diff
@@ -0,0 +1,72 @@
+    }
+    private fun bottomNavScrollToTop(){
+        binding.mainBtmNavigation.setOnItemReselectedListener {
+            if(it.itemId==R.id.btm_home_menu){
```
```suggestion
            if(it.itemId == R.id.btm_home_menu){
```
리포맷 코드가 적용해주시면 좋을 것 같네용!

youngjinc commented on app/src/main/java/org/sopt/sample/presentation/MainActivity.kt:
```diff
@@ -0,0 +1,72 @@
+                        .replace(binding.mainFragmentFrame.id, SearchFragment.newInstance())
+                        .commit()
+                    true
+                }
```
버튼을 누를 때마다 프래그먼트 객체를 직접 생성하는 것보다는 Fragment 재생성 시 알아서 잘 만들 수 있도록 수정해보면 좋겠네용 😉 [공식 문서](https://developer.android.com/guide/fragments/transactions)에 따르면 Fragment 객체를 사용하는 것보다 클래스로 관리하는 것을 권장하고 있어요! 예시는 다음과 같습니다~
```kotlin
supportFragmentManager.commit {
       replace<ExampleFragment>(R.id.fragment_container)
}
```

<img width="887" alt="image" src="https://user-images.githubusercontent.com/48701368/196008373-de0bb060-42f3-463c-b36b-29880ac5d277.png">

그리고 true는 when절 가장 하단으로 빼는 것이 좋겠습니다! 각 케이스마다 중복되니까요~!

youngjinc commented on app/src/main/java/org/sopt/sample/presentation/gallery/GalleryFragment.kt:
```diff
@@ -0,0 +1,44 @@
+
+    private val dataList = listOf<GalleryRecycleData>(
+        GalleryRecycleData(R.drawable.gallery_recycle_ironman1,"아이언맨1"),
+        GalleryRecycleData(R.drawable.gallery_recycle_ironman3,"아이언3"),
+        GalleryRecycleData(R.drawable.gallery_recycle_spiderman_amazing,"어메이징 스파이더맨"),
+        GalleryRecycleData(R.drawable.gallery_recycle_spiderman_farfromhome,"스파이더맨 파프롬홈"),
+        GalleryRecycleData(R.drawable.gallery_recycle_spiderman_nowayhome,"스파이더맨 노웨이홈"),
+        GalleryRecycleData(R.drawable.gallery_recycle_avengers_endgame,"어벤저스 엔드게임"),
```
오호라 마블을 좋아하시는군요~!

hwangsunwoong replied:
😮👍👍

youngjinc commented on app/src/main/java/org/sopt/sample/presentation/gallery/GalleryFragment.kt:
```diff
@@ -0,0 +1,44 @@
+
+class GalleryFragment() : BindingFragment<FragmentGalleryBinding>(FragmentGalleryBinding::bind,R.layout.fragment_gallery) {
+
+    private val dataList = listOf<GalleryRecycleData>(
```
```suggestion
    private val dataList = listOf(
```
담으신 마블 데이터(GalleryRecycleData)로 타입추론이 가능하기 때문에 생략도 가능합니당

youngjinc commented on app/src/main/java/org/sopt/sample/presentation/gallery/data/GalleryRecycleData.kt:
```diff
@@ -0,0 +1,6 @@
+data class GalleryRecycleData(
+    val galleryImg:Int,
+    val desc:String
+)
```
여기도 리포맷 코드 적용이 필요헤보이네요~!

youngjinc commented on app/src/main/java/org/sopt/sample/presentation/gallery/diffUtilAdapter/GalleryDiffUtilItemCallback.kt:
```diff
@@ -0,0 +1,19 @@
+    }
+
+
+}
```
오 DiffUtilItemCallback 이군요! 아주 좋습니다~~
GalleryRecycleData -> T 제네릭 타입으로 변경하여 GalleryRecycleData 뿐만 아니라 추후 다른 데이터 클래스에 대해서도 해당 클래스를 사용해볼 수 있을 것 같아요~! 그렇게 된다면 데이터 클래스 별로 모두 `~~DiffUtilItemCallback` 클래스 파일을 작성하지 않아도 됩니다!

Larry7939 replied:
제네릭을 잊고 있었는데.. 사실은 매번 만들 필요가 없었네요 ㅎㅎ 감사합니다!

youngjinc commented on app/src/main/java/org/sopt/sample/presentation/gallery/diffUtilAdapter/GalleryListAdapter.kt:
```diff
@@ -0,0 +1,39 @@
+
+class GalleryListAdapter(context: Context):ListAdapter<GalleryRecycleData,GalleryViewHolder>(GalleryDiffUtilItemCallback()) {
+    private val inflater by lazy { LayoutInflater.from(context) } //by laze : 초기화를 최대한 늦추는 효과
+    private var dataList:List<GalleryRecycleData> = emptyList()
```
리스트 어뎁터를 사용하신다면1 데이터 리스트를 들고 있지 않아도 된답니다! 아래 onBindViewHolder에서 작성하신 것 처럼 currentList로 바로 접근이 가능하죠~

youngjinc commented on app/src/main/java/org/sopt/sample/presentation/gallery/diffUtilAdapter/GalleryListAdapter.kt:
```diff
@@ -0,0 +1,39 @@
+
+    override fun onBindViewHolder(holder: GalleryViewHolder, position: Int) {
+        when(holder){
+            is GalleryImgViewHolder -> holder.onBind(currentList[position])
```
Gooooood!

youngjinc commented on app/src/main/java/org/sopt/sample/presentation/gallery/diffUtilAdapter/GalleryListAdapter.kt:
```diff
@@ -0,0 +1,39 @@
+
+    override fun getItemCount(): Int {
+        return currentList.size
+    }
```
리스트 어뎁터를 사용하신다면2 또한 getItemCount를 오버라이드 할 필요가 없다면! 해당 함수는 없어도 됩니다~

youngjinc commented on app/src/main/java/org/sopt/sample/presentation/gallery/diffUtilAdapter/GalleryListAdapter.kt:
```diff
@@ -0,0 +1,39 @@
+    fun setData(dataList:List<GalleryRecycleData>){
+        val newData = mutableListOf<GalleryRecycleData>()
+        newData.addAll(currentList)
+        newData.add(GalleryRecycleData(R.drawable.gallery_recycle_ironman1,"아이언맨1"))
```
항상 아이언맨1이 최하단에 보이도록 하기 위함인가요?! 위 코드 처럼 작성하신 이유가 궁금합니다!

Larry7939 replied:
add가 제대로 되는지 테스트 해보기 위해 작성한 코드였습니다 😀

youngjinc commented on app/src/main/java/org/sopt/sample/presentation/gallery/diffUtilAdapter/GalleryListAdapter.kt:
```diff
@@ -0,0 +1,39 @@
+        submitList(newData)
+//        this.dataList = dataList.toList()
+//        submitList(this.dataList)
+//        notifyDataSetChanged()
```
추후 사용하지 않거나 참고할 필요가 없다면 해당 주석은 정리해도 좋겠네요~!

youngjinc commented on app/src/main/java/org/sopt/sample/presentation/gallery/viewholder/GalleryViewHolder.kt:
```diff
@@ -0,0 +1,15 @@
+import org.sopt.sample.databinding.GalleryImgItemViewBinding
+import org.sopt.sample.presentation.gallery.data.GalleryRecycleData
+
+abstract class GalleryViewHolder(private val binding:GalleryImgItemViewBinding):RecyclerView.ViewHolder(binding.root) {
+    abstract fun onBind(data: GalleryRecycleData)
+}
```
갤러리 뷰홀더의 다양한 형태를 위해 추상클래스로 만드신 걸까요?!

Larry7939 replied:
네 맞습니다! ImgViewHolder뿐만 아니라 다른 뷰홀더도 추가할 생각입니다😄

youngjinc commented on app/src/main/java/org/sopt/sample/presentation/home/HomeFragment.kt:
```diff
@@ -0,0 +1,111 @@
+        }
+    }
+
+    companion object {
+        fun newInstance(): HomeFragment {
+            return HomeFragment()
+        }
+    }
```
트랜젝션 관련 코멘트를 반영하시면 해당 코드는 필요없게 됩니다 :)

Larry7939 replied:
앗 그렇네요! 매번 새로 만들 필요가 없으니 삭제해도 되겠네요~

youngjinc commented on app/src/main/java/org/sopt/sample/presentation/home/data/HomeRecycleData.kt:
```diff
@@ -0,0 +1,21 @@
+package org.sopt.sample.presentation.home.data
+
+
+abstract class HomeRecycleData {
+    companion object HomeConst{
+        const val REPO_TITLE_TYPE = 0
+        const val REPO_CONTENT_TYPE = 1
+    }
+    abstract val viewType:Int
+}
+data class HomeRepoTitleData(
+    val titleName: String,
+    override val viewType: Int = REPO_TITLE_TYPE
+): HomeRecycleData()
+
+data class HomeRepoContentData(
+    val profileImg: Int,
+    val repoName: String,
+    val authorName: String,
+    override val viewType: Int = REPO_CONTENT_TYPE
+): HomeRecycleData()
```
공통으로 사용할 뷰타입을 위해서 추상 클래스로 만드신 것일까요?! 해당 방법도 좋습니다 :)
실드클래스를 사용해볼 수도 있을 것 같습니다! 뷰타입 변수도 따로 필요없을 것 같고 타입 체크로 뷰 타입을 구분해볼 수도 있겠어요~

Larry7939 replied:
실드 클래스 너무 좋네요~! enum class의 확장판이군요😀😀

youngjinc commented on app/src/main/java/org/sopt/sample/presentation/home/diffUtilAdapter/HomeRepoListAdapter.kt:
```diff
@@ -0,0 +1,102 @@
+        }
+    }
+
+    //dataList 요소들의 viewType 반환
+    override fun getItemViewType(position: Int): Int {
+        return currentList[position].viewType
+    }
```
Gooooood!

youngjinc commented on app/src/main/java/org/sopt/sample/presentation/home/diffUtilAdapter/HomeRepoListAdapter.kt:
```diff
@@ -0,0 +1,102 @@
+        return currentList.size
+    }
+
+    fun setData(dataList: List<HomeRecycleData>) {
```
리스트 어뎁터를 사용한다면 어뎁터 내에 setData와 같은 데이터 초기화 함수를 만들지 않고 외부 클래스에서 adapter.submitList()를 바로 호출할수도 있습니다~!

youngjinc commented on app/src/main/java/org/sopt/sample/presentation/home/diffUtilAdapter/HomeRepoListAdapter.kt:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    fun add() {
+        val newData = mutableListOf<HomeRecycleData>()
+        newData.addAll(currentList)
+        newData.add(
+            HomeRepoContentData(
+                R.drawable.home_profile_img_bino2,
+                "NEW IN_SOPT",
+                "Jihun Park",
+                REPO_CONTENT_TYPE
+            ) as HomeRecycleData
+        )
+        submitList(newData)
```
kotlin.collections.List에서의 plus 연산자는 오버라이드되어있기 때문에 아래와 같이 표현해볼 수도 있겠습니다~
```suggestion
        val data = currentList + (HomeRepoContentData(
                R.drawable.home_profile_img_bino2,
                "NEW IN_SOPT",
                "Jihun Park",
                REPO_CONTENT_TYPE
            ) as HomeRecycleData)
        submitList(data.toMutableList())
```

youngjinc commented on app/src/main/java/org/sopt/sample/presentation/home/viewholder/HomeViewHolder.kt:
```diff
@@ -0,0 +1,35 @@
+class RepoTitleViewHolder(private val binding: HomeRepoTitleItemViewBinding) : HomeViewHolder(binding){
+    override fun onBind(data: HomeRecycleData) {
+        val titleData = data as HomeRepoTitleData
+        binding.apply {
+            homeItemTitle.text = titleData.titleName
+        }
```
binding객체 내의 homeItemTitle의 text속성 하나만 변경하는 것이라면 
`binding.homeItemTitle.text = titleData.titleName` 으로 작성하는 것이 더 간단해보입니당!

youngjinc commented on app/src/main/java/org/sopt/sample/presentation/home/viewholder/HomeViewHolder.kt:
```diff
@@ -0,0 +1,35 @@
+    //2. onCreateViewHolder에서 받아온 binding을 이용하여 HomeItemView의 컴포넌트에 접근할 수 있다.
+    override fun onBind(data: HomeRecycleData) {
+        val contentData = data as HomeRepoContentData
+        binding.apply {
+            homeItemProfile.setImageDrawable(binding.root.context.getDrawable(contentData.profileImg))
+            homeItemRepoName.text = contentData.repoName
+            homeItemAuthor.text = contentData.authorName
+        }
```
apply를 통해 반환된 바인딩 객체 자체를 어딘가에서 사용할 것이 아니라면 `with(binding) {}` 사용도 좋아보입니다~

youngjinc reviewed (approved):
리뷰 완! getItemViewType 오버라이드 잘하셨네용!! 굉장하십니다~ ListAdapter 사용도 아주 좋습니다~ 고생 많으셨습니다 :)

KWY0218 commented on app/src/main/java/org/sopt/sample/presentation/MainActivity.kt:
```diff
@@ -0,0 +1,72 @@
+    private fun bottomNavScrollToTop(){
+        binding.mainBtmNavigation.setOnItemReselectedListener {
+            if(it.itemId==R.id.btm_home_menu){
+                findViewById<RecyclerView>(R.id.home_recycler_repo).smoothScrollToPosition(0)
```
findViewById 말고 binding 사용하면 좋을 것 같습니다

youngjinc replied:
오호라.. findViewById가 있었군요👀

KWY0218 commented on app/src/main/java/org/sopt/sample/presentation/login/SignInActivity.kt:
```diff
@@ -51,19 +51,22 @@ class SignInActivity : BindingActivity<ActivitySigninBinding>(ActivitySigninBind
                     }
 
-                    binding.root.makeSnackBar(getString(R.string.signin_success_signup)).setAnchorView(binding.signInLoginBtn)
+                    binding.root.makeSnackBar(getString(R.string.signin_success_signup))
+                        .setAnchorView(binding.signInLoginBtn)
                         .show()
                 }
```
`SignInActivity` 코드를 봤을 때 cordinator Layout을 사용한 것 같은데, 혹시 스낵바가 생겼을 때 다른 위젯들이 스낵바에 영향을 받아서 위치가 조정이 되나요? 

아니라면
 상관 없지만 만약 위치가 조정이 된다면

[Material Desgin](https://material.io/components/snackbars#behavior) Consecutive snackbars 부분을 보시면
Don’t animate other components along with snackbar animations, such as the floating action button.
라고 나와 있습니다.

이 부분 유의해서 스낵바가 나타날 위치를 고려하면 좋을 것 같습니다.

sujin-kk commented on app/src/main/java/org/sopt/sample/base/BindingFragment.kt:
```diff
@@ -0,0 +1,32 @@
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
```
`View `로 리턴해도 되지 않을까요?!

sujin-kk commented on app/src/main/java/org/sopt/sample/presentation/gallery/GalleryFragment.kt:
```diff
@@ -0,0 +1,44 @@
+class GalleryFragment() : BindingFragment<FragmentGalleryBinding>(FragmentGalleryBinding::bind,R.layout.fragment_gallery) {
+
+    private val dataList = listOf<GalleryRecycleData>(
+        GalleryRecycleData(R.drawable.gallery_recycle_ironman1,"아이언맨1"),
+        GalleryRecycleData(R.drawable.gallery_recycle_ironman3,"아이언3"),
+        GalleryRecycleData(R.drawable.gallery_recycle_spiderman_amazing,"어메이징 스파이더맨"),
+        GalleryRecycleData(R.drawable.gallery_recycle_spiderman_farfromhome,"스파이더맨 파프롬홈"),
+        GalleryRecycleData(R.drawable.gallery_recycle_spiderman_nowayhome,"스파이더맨 노웨이홈"),
+        GalleryRecycleData(R.drawable.gallery_recycle_avengers_endgame,"어벤저스 엔드게임"),
```
흥미롭군요,,

sujin-kk commented on app/src/main/java/org/sopt/sample/presentation/home/diffUtilAdapter/HomeRepoListAdapter.kt:
```diff
@@ -0,0 +1,102 @@
+import org.sopt.sample.presentation.home.viewholder.RepoTitleViewHolder
+
+
+class HomeRepoListAdapter(context: Context) :
+    ListAdapter<HomeRecycleData, HomeViewHolder>(HomeRepoDiffUtilItemCallback()) {
+    private val inflater by lazy { LayoutInflater.from(context) } //by laze : 초기화를 최대한 늦추는 효과
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): HomeViewHolder {
+        return when (viewType) {
+            REPO_TITLE_TYPE -> RepoTitleViewHolder(
+                HomeRepoTitleItemViewBinding.inflate(
+                    inflater,
+                    parent,
+                    false
+                )
+            )
+            REPO_CONTENT_TYPE -> RepoContentViewHolder(
+                HomeRepoContentItemViewBinding.inflate(
+                    inflater,
+                    parent,
+                    false
+                )
+            )
+            else -> {
+                throw java.lang.IllegalArgumentException()
+            }
+        }
+    }
+
```
어댑터 생성자에 context를 주어서 inflater를 초기화 하기보다는
LayoutInflater가 쓰이는 onCreateViewHolder 안에서 
`LayoutInfalter.(parent.context)`로 접근해도 좋지 않을까요?!?

sujin-kk commented on app/src/main/java/org/sopt/sample/base/BindingFragment.kt:
```diff
@@ -0,0 +1,32 @@
+import androidx.viewbinding.ViewBinding
+
+// Fragment의 기본을 작성, 뷰 바인딩 활용
+abstract class BindingFragment<B : ViewBinding>(
```
`BindingFragment `네이밍 good

sujin-kk commented on app/src/main/java/org/sopt/sample/presentation/gallery/diffUtilAdapter/GalleryDiffUtilItemCallback.kt:
```diff
@@ -0,0 +1,19 @@
+
+class GalleryDiffUtilItemCallback:DiffUtil.ItemCallback<GalleryRecycleData>() {
+
+    override fun areItemsTheSame(oldItem: GalleryRecycleData, newItem: GalleryRecycleData): Boolean {
+        //oldItem과 newItem의 id 비교
+        return oldItem == newItem
+    }
+
+    override fun areContentsTheSame(oldItem: GalleryRecycleData, newItem: GalleryRecycleData): Boolean {
+        //oldItem과 newItem의 내용 비교
+        return oldItem.equals(newItem)
+    }
```
코틀린에서 ==와 .equals는 동일한 값을 가지는지 체크하는 연산자이므로  두 함수가 같은 기능을 하는 것으로 보입니다!

보통 areItemsTheSame은 고유값 비교(id, hash 등)를 통해 같은 값을 가지는지를 계산하고
areContentsTheSame은 두 아이템의 값이 정확히 같은지를 비교하게됩니다.
(id가 같아도 다른 값이 업데이트 된다면 정확히 동일한 것은 아니겠죠?)
(==로 가능, ===는 메모리 상 같은 주소값 -> 같은 객체인지 비교하는 것)

sujin-kk reviewed (approved):
수고하셨어요~~👍🏻

hwangsunwoong commented on app/src/main/java/org/sopt/sample/base/BindingFragment.kt:
```diff
@@ -0,0 +1,32 @@
+    @LayoutRes layoutResId: Int
+) : Fragment(layoutResId) {
+    private var _binding: B? = null
+    protected val binding
+        get() = requireNotNull(_binding!!)
```
```suggestion
    protected val binding
        get() = requireNotNull(_binding!!) { "${this::class.java.simpleName} error." }
```
예외 처리를 추가해도 좋을 것 같아요 ~!

hwangsunwoong commented on app/src/main/java/org/sopt/sample/presentation/home/diffUtilAdapter/HomeRepoListAdapter.kt:
```diff
@@ -0,0 +1,102 @@
+                )
+            )
+            else -> {
+                throw java.lang.IllegalArgumentException()
```
```suggestion
                throw IllegalArgumentException()
```
kotlin exception을 활용하면 좋을 것 같아요!

hwangsunwoong commented on app/src/main/java/org/sopt/sample/util/const/UserInfo.kt:
```diff
@@ -0,0 +1,5 @@
+package org.sopt.sample.util.const
+
+const val USER_INFO_ID = "id"
+const val USER_INFO_PW = "pw"
+const val USER_INFO_MBTI = "mbti"
```
👍👍

hwangsunwoong commented on app/src/main/res/layout/fragment_home.xml:
```diff
@@ -0,0 +1,46 @@
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    tools:context=".presentation.home.HomeFragment">
+
+    <Button
+        android:id="@+id/home_recycler_repo_add_btn"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:backgroundTint="@color/black"
+        android:text="@string/home_repo_add_btn">
+    </Button>
+    <Button
+        android:id="@+id/home_recycler_repo_remove_btn"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:backgroundTint="@color/black"
+        android:text="@string/home_repo_remove_btn">
+    </Button>
+    <androidx.constraintlayout.helper.widget.Flow
+        android:id="@+id/home_btn_flow"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        app:layout_constraintTop_toTopOf="parent"
+        android:layout_marginTop="50dp"
+        android:orientation="horizontal"
+        app:constraint_referenced_ids="home_recycler_repo_add_btn,home_recycler_repo_remove_btn"
+        >
+    </androidx.constraintlayout.helper.widget.Flow>
+    <androidx.recyclerview.widget.RecyclerView
+        android:id="@+id/home_recycler_repo"
+        android:layout_width="match_parent"
+        android:layout_height="0dp"
+        android:layout_marginTop="40dp"
+        app:layout_constraintTop_toBottomOf="@id/home_btn_flow"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layoutManager="androidx.recyclerview.widget.LinearLayoutManager"
+        tools:listitem="@layout/home_repo_content_item_view"
+        >
+    </androidx.recyclerview.widget.RecyclerView>
```
자식 뷰가 없는 경우 닫는 괄호를 "/>"로 바꿔주시면 좋을 것 같아요!

hwangsunwoong reviewed (approved):
2주차 과제 고생 많으셨어요 ~~ 😊

Larry7939 committed:
add : HomeFragmenr 레포지터리 add 시 scrollToTop 기능

MainActivity에서 싱글톤으로 mContext 초기화 한 후,
HomeRepoListAdapter에서 mContext를 이용하여 HomeFragment의 scrollToTop 메소드에 접근

Larry7939 committed:
chore

Larry7939 committed:
chore : reformat code 및 DiffUtilItemCallback을 제네릭 클래스로 변경

HomeRepoListAdapter와 GalleryListAdapter에서 따로 쓰던 DiffUtilItemCallback을 제네릭 클래스로 변경 후, util폴더로 이동
이후 ListAdapter가 추가된다 하더라도, 새로운 DiffUtilItemCallback의 추가 없이, util폴더의 DiffUtilItemCallback을 사용.

Larry7939 committed:
chore : GalleryListAdapter 수정

dataList삭제 -> 현재 리스트 데이터는 currentList로 접근 가능

Larry7939 committed:
chore : ListAdapter getItemCount 삭제

ListAdapter 사용 시, getItemCount 필요없음!

Larry7939 committed:
chore : HomeFragment

MainActivity에서 Fragment Transaction관련 코드 수정으로, 매번 직접 새로운 Fragment를 만들어내지 않기 때문에 newInstance 코드 삭제

Larry7939 committed:
chore : HomeRecycleData에 Sealed class적용

viewType변수가 삭제되었기 때문에,
HomeRepoListAdapter에서도 viewType이 아닌 타입 체크로 뷰타입 구분

Larry7939 committed:
chore : HomeRepoListAdapter setData 메소드 삭제

setData 메소드에서 호출한 submitList는 HomeFragment에서 바로 호출 가능

Larry7939 committed:
chore : 불필요한 주석 삭제

Larry7939 committed:
chore : DiffUtilItemCallback의 동등성 검사 수정

Larry7939 committed:
chore : BindingFragment의 예외 처리 추가

Larry7939 committed:
chore : 예외처리 Kotlin exception으로 대체

HomeRepoListAdapter, MainActivity에 적용된 사항

Larry7939 committed:
chore : 자식 뷰가 없는 경우에 대해 닫는 괄호 "/>"로 변경

Larry7939 commented:
코드 리뷰 반영 완료
